### PR TITLE
- IdentityServer4, Admin, Admin.Api, STS.Identity, added support for environment-specific configuration files

### PR DESCRIPTION
**Pull request by ViRuSTriNiTy**
… in logging setup and `CreateHostBuilder()`, see issue #549

### CODE REVIEW

File: src/Skoruba.IdentityServer4.Admin.Api/Program.cs [hunk 1]
```diff
@@ -38,7 +38,9 @@ private static IConfiguration GetConfiguration(string[] args)
             var configurationBuilder = new ConfigurationBuilder()
                 .SetBasePath(Directory.GetCurrentDirectory())
                 .AddJsonFile("appsettings.json", optional: false, reloadOnChange: true)
-                .AddJsonFile("serilog.json", optional: true, reloadOnChange: true);
+                .AddJsonFile($"appsettings.{environment}.json", optional: true, reloadOnChange: true)
```
Comment (b0) [inline src/Skoruba.IdentityServer4.Admin.Api/Program.cs]: Do we really need this line ? I think apps already works with appsettings{environemnt}.json by default.

File: src/Skoruba.IdentityServer4.Admin/Program.cs [hunk 2]
```diff
@@ -61,7 +61,9 @@ private static IConfiguration GetConfiguration(string[] args)
             var configurationBuilder = new ConfigurationBuilder()
                 .SetBasePath(Directory.GetCurrentDirectory())
                 .AddJsonFile("appsettings.json", optional: false, reloadOnChange: true)
-                .AddJsonFile("serilog.json", optional: true, reloadOnChange: true);
+                .AddJsonFile($"appsettings.{environment}.json", optional: true, reloadOnChange: true)
```
Comment (b0) [inline src/Skoruba.IdentityServer4.Admin/Program.cs]: Do we really need this line ? I think apps already works with appsettings{environemnt}.json by default.

File: src/Skoruba.IdentityServer4.Admin.Api/Program.cs (hunk 1 above)
Comment (ViRuSTriNiTy) [reply src/Skoruba.IdentityServer4.Admin.Api/Program.cs]: Imho yes, this line is necessary as there is no "automagic" that is going on here, see https://github.com/skoruba/IdentityServer4.Admin/issues/549#issuecomment-603730495

File: src/Skoruba.IdentityServer4.Admin/Program.cs (hunk 2 above)
Comment (ViRuSTriNiTy) [reply src/Skoruba.IdentityServer4.Admin/Program.cs]: Imho yes, see https://github.com/skoruba/IdentityServer4.Admin/pull/551/files#r397750808

**Comment by b0** [comment]
At least how I am reading this: https://docs.microsoft.com/en-us/aspnet/core/fundamentals/configuration/?view=aspnetcore-3.1#default-configuration

appsettings.json and appsettings.{environment}.json are "magicaly" provided by default. 

Best regards

**Comment by ViRuSTriNiTy** [comment]
@bo Yes i read this section too but i think its a bit misleading. I've checked the .NET Core sources and as i mentioned in the related issue the `appsettings.{environment}.json` is added by the `DefaultBuilder`, see https://github.com/dotnet/aspnetcore/blob/master/src/DefaultBuilder/src/WebHost.cs#L170. The thing here is that the logging configuration is created without the `DefaultBuilder` hence you need to do it manually.

**Comment by b0** [comment]
Hi @ViRuSTriNiTy ,

indeed, logging configuration is created differently and those lines are needed. 

Great work. 👍 

Thank you

**Comment by ViRuSTriNiTy** [comment]
@b0 Thanks for reviewing. How to remove the merge block? Is @skoruba needed to resolve this?

**Comment by ViRuSTriNiTy** [comment]
@skoruba Please unblock the PR when everything is ok.

File: src/Skoruba.IdentityServer4.Admin.Api/Program.cs (hunk 1 above)
Comment (ekjuanrejon) [reply src/Skoruba.IdentityServer4.Admin.Api/Program.cs:41]: I don't think we need this line. If anyone wants a custom serilog configuration. The configuration should be mounted into the application. 

Define in your release pipeline to mount the serilog.json from the outside to the inside. The proposed line will not work with docker and kubernetes

Comment (ViRuSTriNiTy) [reply src/Skoruba.IdentityServer4.Admin.Api/Program.cs:41]: > I don't think we need this line. If anyone wants a custom serilog configuration. The configuration should be mounted into the application.
> 
> Define in your release pipeline to mount the serilog.json from the outside to the inside. The proposed line will not work with docker and kubernetes

Thanks for your feedback but i don't understand your instructions. What does "mounted into the application" actually mean and what "release pipeline"? Do you mean i should copy Serilog.Development.json to Serilog.json before releasing the application?

Comment (ekjuanrejon) [reply src/Skoruba.IdentityServer4.Admin.Api/Program.cs:41]: if you look at the docker compose below
https://github.com/skoruba/IdentityServer4.Admin/blob/master/docker-compose.yml#L35

This give the flexibility for anyone to define serilog.json as they want and mount it from the host machine to the container

Comment (ekjuanrejon) [reply src/Skoruba.IdentityServer4.Admin.Api/Program.cs:41]: In docker or kubernetes if you have multiple applications that required serilog you only want to have one config. In kubernetes you can use a config map or secret and mount those inside the pod. 

If it need to change you change the config map without having to redploy your application

Comment (ViRuSTriNiTy) [reply src/Skoruba.IdentityServer4.Admin.Api/Program.cs:41]: @ekjuanrejon So this PR is not necessary at all?

Comment (ekjuanrejon) [reply src/Skoruba.IdentityServer4.Admin.Api/Program.cs:41]: To me it is not necessary but if you want I'd say if you still want to do it. do it in your fork

Comment (ViRuSTriNiTy) [reply src/Skoruba.IdentityServer4.Admin.Api/Program.cs:41]: So you want only "necessary" changes not improvements, that's sad as my approach also [allows config transformations](https://stackoverflow.com/questions/53208920/equivalent-to-app-config-transforms-for-net-core) via [SlowCheetah](https://marketplace.visualstudio.com/items?itemName=vscps.SlowCheetah-XMLTransforms), all effort in vain. Closing the PR.

**Comment by ViRuSTriNiTy** [comment]
PR closed as a result of https://github.com/skoruba/IdentityServer4.Admin/pull/551#discussion_r407981004

File: src/Skoruba.IdentityServer4.Admin.Api/Program.cs (hunk 1 above)
Comment (ekjuanrejon) [reply src/Skoruba.IdentityServer4.Admin.Api/Program.cs:41]: I am not the owner of the repo. I just gave a comment.

Comment (ekjuanrejon) [reply src/Skoruba.IdentityServer4.Admin.Api/Program.cs:41]: The owner can very well decide to approve your pr

Comment (ekjuanrejon) [reply src/Skoruba.IdentityServer4.Admin.Api/Program.cs:41]: @skoruba

**Comment by skoruba** [comment]
@ViRuSTriNiTy - please, reopen this PR. I want to use it.
Thanks.

**Comment by ViRuSTriNiTy** [comment]
@skoruba Reopened as requested

**Comment by skoruba** [comment]
Thanks 👍🏼

**Pull request merged**